src/main/scala/viper/server/vsi/Cache.scala

```diff
@@ -0,0 +1,204 @@
+      case None =>
+        //if file not in cache yet, create new map entry for it. Recall the function.
+        _cache += (file_key -> collection.mutable.Map[String, List[CacheEntry]]())
+        update(file_key, key, dependencies, content)
+    }
+  }
+
+  def forgetFile(file_key: String): Option[String] = {
+    _cache.remove(file_key) match {
+      case Some(_) => Some(file_key)
+      case None => None
+    }
+  }
+
+  def resetCache(): Unit = {
+    _cache.clear()
+  }
+}
+
+// ===== AUXILIARY TRAITS ==================================================================
+
+/** This trait is a generic wrapper for cache entries.
+  *
+  * Extending this specifies
+  *
+  *  - What the cache will be holding.
+  *  - Which of the AST member it will be holding it for
+  *  - The dependency hash of when the entry is stored.
+  *
+  *  The dependency hash reflects the state of the dependencies at the time when the entry was created.
+  *  If at some point a members dependency hash is no longer equal to the one stored here, the
+  *  entry is no longer valid.
+  * */
+class CacheEntry(val concerning: Concerning, val cacheContent: CacheContent, val depencyHash: String) {
+}
+
+
+trait CacheContent {
+}
+
+/** This trait is a generic wrapper for an AST.
+  *
+  * An AST might a number of members for which it might be sensible to store results, errors, etc.
+  * In order for the cache to know which these are, an AST must implement the de-/compose methods.
+  *
+  * These should decompose and AST into a list of all the members for which the cache is to store
+  * results and compose an AST given such list of members.
+  * */
+trait AST {
+  def compose(cs: List[Concerning]): AST
+
+  def decompose(): List[Concerning]
+}
+
+trait Member {
+  def hashFunction(): String
+}
+
+/** This trait is a generic wrapper for AST Members for which things are stored in the cache.
+  *
+  * An AST might a number of members for which it might be sensible to store results, errors, etc.
+  * However, there is a set of requirements of any member for which cache is used. This comes in
+  * the form of the following three methods
+  *
+  * hashFunction: A member must be hashable to determined its mapping in the cache.
+  *
+  * transform: A member that has been retrieved from the cache should not be reverified. Therefore,
+  * a member must have an alternative representation that allows the verifier to distinguish if from
+  * members that missed the cache and need reverification
+  *
+  * getDependencies: A member may hit the cache, but the attached verification results might be invalid.
+  * Reason for this is that other members in the program have changed in such a way that influences the
+  * current member. These influencing members are called dependencies and need to be checked when
+  * retrieving a member's attached result from cache.
+  * */
+trait Concerning extends Member {
+  def transform: Concerning
+
+  def getDependencies(ast: AST): List[Member]
```

Comment:
It's a bit weird that this is not a method of class ```AST```. Is there a good reason for that?